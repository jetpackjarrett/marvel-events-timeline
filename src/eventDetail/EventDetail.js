import React, { Component, PropTypes } from 'react';
import { fetchFromMarvel, getImageUrl } from '../services/marvelApi';
import Creators from './Creators';
import Comics from './Comics';
import './eventDetail.css';

export default class EventDetail extends Component {

  static propTypes = {
    comics: PropTypes.shape({
      collectionURI: PropTypes.string.isRequired,
    }).isRequired,
    creators: PropTypes.shape({
      items: PropTypes.array.isRequired,
    }).isRequired,
    description: PropTypes.string,
    onClose: PropTypes.func.isRequired,
    thumbnail: PropTypes.shape({
      extension: PropTypes.string.isRequired,
      path: PropTypes.string.isRequired,
    }).isRequired,
    title: PropTypes.string.isRequired,
    urls: PropTypes.array.isRequired,
  };

  state = {};

  componentDidMount() {
    fetchFromMarvel(this.props.comics.collectionURI)
      .then((response) => this.setState({ comics: response.data.results }));
  }

  render() {
    const { urls, description, creators, title, onClose } = this.props;
    const { comics } = this.state;
    const image = getImageUrl(this.props.thumbnail, 'standard_fantastic');

    return (
      <div className="event-detail-container">
        <button onClick={onClose} className="event-detail-close-button">
          &times;
        </button>
        <div className="event-detail">
          <header className="event-detail-header">
            <div><img src={image} alt="cover" /></div>
            <div className="event-detail-header-text">
              <h3>{title}</h3>
              <p>{description}</p>
              {urls.map(({ url, type }, i) => (
                <a
                  key={i}
                  target="_blank"
                  rel="noopener noreferrer"
                  href={url}>
                  {type}
                </a>
              ))}
            </div>
          </header>
          <div className="event-detail-content">
            <Creators creators={creators.items} />
            <Comics comics={comics} />
          </div>
        </div>
      </div>
    );
  }
}
