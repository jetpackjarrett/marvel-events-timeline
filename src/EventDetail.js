import React, { Component } from 'react';
import { fetchFromMarvel, getImageUrl } from './services/marvelApi';
import './eventDetail.css';

const Creators = ({ creators }) => {
  const creatorsByRole = creators.reduce((memo, { role, name }) => {
    if (!memo[role]) {
      memo[role] = [];
    }

    memo[role].push(name);
    return memo;
  }, {});

  return (
    <section>
      <h4>Creators</h4>
      <ul className="event-detail-header-text-creators">
        {Object.keys(creatorsByRole).map(role => {
          return (
            <li key={role} className="event-detail-header-text-creator">
              <strong>{role}</strong>
              <span>{creatorsByRole[role].join(', ')}</span>
            </li>
          );
        })}
      </ul>
    </section>
  );
};

const Comics = ({ comics = [] }) => {
  return (
    <section>
      <h4>Comics</h4>
      <div className="event-detail-comics">
        {comics.length === 0 && <div className="loader" />}
        {comics
          .filter(comic => comic.issueNumber > 0)
          .map(({ id, thumbnail, title, comic, urls }) => {

          return (
            <div key={id} className="event-detail-comic">
              <a target="_blank" href={urls[0].url}>
                <img src={getImageUrl(thumbnail, 'portrait_medium')} alt={title} />
              </a>
            </div>
          );
        })}
      </div>
    </section>
  );
};

export default class EventDetail extends Component {

  state = {};

  componentDidMount() {
    fetchFromMarvel(this.props.comics.collectionURI)
      .then(response => this.setState({ comics: response.data.results }));
  }

  render() {
    const { urls, description, creators, title, onClose } = this.props;
    const { comics } = this.state;
    const image = getImageUrl(this.props.thumbnail, 'standard_fantastic');

    return (
      <div className="event-detail-container">
        <button onClick={onClose} className="event-detail-close-button">
          &times;
        </button>
        <div className="event-detail">
          <header className="event-detail-header">
            <div><img src={image} alt="cover" /></div>
            <div className="event-detail-header-text">
              <h3>{title}</h3>
              <p>{description}</p>
              {urls.map((url, i) => <a key={i} target="_blank" href={url.url}>{url.type}</a>)}
            </div>
          </header>
          <div className="event-detail-content">
            <Creators creators={creators.items} />
            <Comics comics={comics} />
          </div>
        </div>
      </div>
    );
  }
}
