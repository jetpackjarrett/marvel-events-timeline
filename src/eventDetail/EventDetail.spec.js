/*eslint import/no-extraneous-dependencies: ["error", { "devDependencies": true }] */
import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import EventDetail from './EventDetail';

jest.mock('../services/marvelApi', () => {
  return {
    getImageUrl: jest.fn(),
    fetchFromMarvel: jest.fn(() => Promise.resolve())
  };
});

describe('EventDetail', () => {
  it('renders the event detail', () => {
    const onClose = jest.fn();
    const component = mount(
      <EventDetail
        comics={{
          collectionURI: 'http://fake-collection-uri.com/comics'
        }}
        creators={{
          items: []
        }}
        description="mock description"
        thumbnail={{
          path: 'http://fake-image.com/large',
          extension: 'jpg',
        }}
        title="Mock Comic"
        urls={[{ type: 'detail', url: 'http://fake-detail.com' }]}
        onClose={onClose} />
    );
    expect(toJson(component)).toMatchSnapshot();
  });
});
