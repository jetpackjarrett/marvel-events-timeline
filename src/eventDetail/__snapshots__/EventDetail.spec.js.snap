exports[`EventDetail renders the event detail 1`] = `
<EventDetail
  comics={
    Object {
      "collectionURI": "http://fake-collection-uri.com/comics",
    }
  }
  creators={
    Object {
      "items": Array [],
    }
  }
  description="mock description"
  onClose={[Function]}
  thumbnail={
    Object {
      "extension": "jpg",
      "path": "http://fake-image.com/large",
    }
  }
  title="Mock Comic"
  urls={
    Array [
      Object {
        "type": "detail",
        "url": "http://fake-detail.com",
      },
    ]
  }>
  <div
    className="event-detail-container">
    <div
      className="event-detail-close-button"
      onClick={[Function]}>
      <span>
        Ã—
      </span>
    </div>
    <div
      className="event-detail">
      <header
        className="event-detail-header">
        <div>
          <img
            alt="cover" />
        </div>
        <div
          className="event-detail-header-text">
          <h3>
            Mock Comic
          </h3>
          <p>
            mock description
          </p>
          <a
            href="http://fake-detail.com"
            rel="noopener noreferrer"
            target="_blank">
            detail
          </a>
        </div>
      </header>
      <div
        className="event-detail-content">
        <Component
          creators={Array []}>
          <section>
            <h4>
              Creators
            </h4>
            <ul
              className="event-detail-creators" />
          </section>
        </Component>
        <Component>
          <section>
            <h4>
              Comics
            </h4>
            <div
              className="event-detail-comics">
              <div
                className="loader" />
            </div>
          </section>
        </Component>
      </div>
    </div>
  </div>
</EventDetail>
`;
